### PR TITLE
[GLUTEN-1476][VL] Enable scan on struct and map types

### PR DESCRIPTION
**Pull request by rui-mo**
## What changes were proposed in this pull request?

Scan on struct and map types are enabled with below limitations:

- Parquet scan of nested map with struct as key type is not supported in Velox (fallback).
- Merge schema option is not supported (fallback).
- Schema pruning is not supported in Velox struct reader (tests disabled).

Depends on: https://github.com/oap-project/velox/pull/290.

\#1476

## How was this patch tested?

Testing on CI.

### CODE REVIEW

**Review by PHILO-HE — APPROVED** [review]

File: gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala [hunk 1]
```diff
@@ -816,15 +824,128 @@ class VeloxTestSettings extends BackendTestSettings {
...
+    .exclude("SPARK-10301 requested schema clipping - schemas overlap but don't contain each other")
+    .exclude("SPARK-10301 requested schema clipping - deeply nested struct")
   // requires resource files from Vanilla spark jar
   // enableSuite[GlutenParquetRebaseDatetimeV1Suite]
   // enableSuite[GlutenParquetRebaseDatetimeV2Suite]
   enableSuite[GlutenParquetV1SchemaPruningSuite]
     // spark.sql.parquet.enableVectorizedReader=true not supported
     .excludeByPrefix("Spark vectorized reader - ")
+    // Struct reader does not support implicit schema pruning.
+    .exclude("Non-vectorized reader - without partition data column - " +
```
Comment (PHILO-HE) [inline gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:839]: As so many tests need to be excluded, can we just use a white list for it if white list size is much smaller? I.e., just include the workable tests.

File: gluten-core/src/main/scala/io/glutenproject/execution/FileSourceScanExecTransformer.scala [hunk 2]
```diff
@@ -123,11 +123,18 @@ class FileSourceScanExecTransformer(@transient relation: HadoopFsRelation,
 
   override def doValidateInternal(): Boolean = {
     // Bucketing table has `bucketId` in filename, should apply this in backends
-    if (!bucketedScan) {
-      super.doValidateInternal()
-    } else {
-      false
+    if (bucketedScan) {
+      logWarning(s"Validation failed for ${this.getClass.toString} due to " +
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/execution/FileSourceScanExecTransformer.scala:127]: @WangGuangxin, please note this change, which looks conflict to your code change in
https://github.com/oap-project/gluten/pull/1865.

File: gluten-ut/spark33/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala [hunk 3]
```diff
@@ -674,12 +682,125 @@ class VeloxTestSettings extends BackendTestSettings {
...
+      "SPARK-38977: schema pruning with correlated IN subquery")
+    .exclude("Non-vectorized reader - without partition data column - " +
+      "SPARK-38977: schema pruning with correlated NOT IN subquery")
+    .exclude("Non-vectorized reader - with partition data column - " +
+      "SPARK-38977: schema pruning with correlated NOT IN subquery")
   enableSuite[GlutenParquetV2SchemaPruningSuite]
     // spark.sql.parquet.enableVectorizedReader=true not supported
     .excludeByPrefix("Spark vectorized reader - ")
+    // Struct reader does not support implicit schema pruning.
+    .exclude("Non-vectorized reader - without partition data column - " +
```
Comment (PHILO-HE) [inline gluten-ut/spark33/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:751]: As above, white list may be better.

File: gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala (hunk 1 above)
Comment (rui-mo) [reply gluten-ut/spark32/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:839]: Although ~30 of them are excluded, there are still ~60 unit tests in this suite, so the list would be smaller with exclude.

**Review by PHILO-HE — APPROVED** [review]
Looks great!

File: gluten-core/src/main/scala/io/glutenproject/execution/FileSourceScanExecTransformer.scala (hunk 2 above)
Comment (jinchengchenghh) [inline gluten-core/src/main/scala/io/glutenproject/execution/FileSourceScanExecTransformer.scala:127]: Please use `logValidateFailure` to track all the failed validation

Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/execution/FileSourceScanExecTransformer.scala]: Changed to use `logValidateFailure`, thanks.

**Review by PHILO-HE — APPROVED** [review]

**Pull request merged**